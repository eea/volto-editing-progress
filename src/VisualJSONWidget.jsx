import React, { useState, useEffect } from 'react';
import { ModalForm } from '@plone/volto/components';
import { JSONSchema } from './schema';
import {
  Button,
  Container,
  Segment,
  Divider,
  Sidebar,
} from 'semantic-ui-react';
import { flattenToAppURL } from '@plone/volto/helpers';
import './less/editor.less';
import _ from 'lodash';
import { FormattedMessage, defineMessages, injectIntl } from 'react-intl';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import { getRawContent } from './actions';
import SidebarComponent from './WidgetSidebar';
import EditDataComponent from './WidgetDataComponent';
import './less/editor.less';
const messages = defineMessages({
  jsonTitle: {
    id: 'Edit JSON',
    defaultMessage: 'Edit JSON',
  },
});

export const SIDEBAR_WIDTH = '250px';
export const COMPONENT_HEIGHT = '750px';

function isValidJson(json) {
  try {
    JSON.parse(json);
    return true;
  } catch (e) {
    return false;
  }
}

function addNewStateToAlreadyExistingField(
  currentContentTypeData,
  currentField,
  statesToAdd,
  message,
) {
  for (
    let localRuleIndex = 0;
    localRuleIndex < currentContentTypeData.length;
    localRuleIndex++
  ) {
    if (currentContentTypeData[localRuleIndex].prefix !== currentField)
      continue;

    // TODO rewrite message as an object with multiple message as a key and add it
    // to currentContentTypeData dinamically in order to create the posibillity
    // for multiple fields

    if (message) currentContentTypeData[localRuleIndex].message = message;
    if (statesToAdd !== undefined) {
      currentContentTypeData[localRuleIndex].states = statesToAdd;
    } else if (!message) currentContentTypeData.splice(localRuleIndex, 1);
  }
}

function doesPrefixExistInCurrentContentTypeData(
  currentContentTypeData,
  currentField,
) {
  return currentContentTypeData.every((rule) => rule.prefix !== currentField);
}

function createFieldRule(currentField, statesToAdd) {
  return {
    prefix: currentField,
    states: statesToAdd,
    condition: 'python:value',
    hideReady: 'False',
    iconEmpty: 'eea-icon eea-icon-edit',
    iconReady: 'eea-icon eea-icon-check',
    labelEmpty: 'Please set the {label} of this {context.portal_type}',
    labelReady: 'You added the {label}',
    link: 'edit#fieldset-supporting information-field-label-data_description',
    linkLabel: 'Add {label}',
    message: '',
  };
}

const VisualJSONWidget = (props) => {
  const { id, value = {}, onChange } = props;
  console.log(value);
  const [isJSONEditorOpen, setIsJSONEditorOpen] = useState(false);
  const [currentContentType, setCurrentContentType] = useState();

  const path = flattenToAppURL(
    currentContentType?.['@id'] ? `${currentContentType['@id']}` : null,
  );

  const dispatch = useDispatch();
  const request = useSelector((state) => state.rawdata?.[path]);
  const content = request?.data;
  const types = useSelector((state) => state.types);

  useEffect(() => {
    if (path && !request?.loading && !request?.loaded && !content)
      dispatch(getRawContent(path));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [dispatch, path, content, request?.loaded, request?.loading]);

  useEffect(() => {
    if (types.loaded && !types.loading && Array.isArray(types.types)) {
      setCurrentContentType(types.types[0]);
    }
  }, [types]);

  const handleOnCancel = (e) => {
    setIsJSONEditorOpen(false);
  };

  const handleEditJSON = (e) => {
    e.preventDefault();
    setIsJSONEditorOpen(true);
  };

  const onJSONSubmit = (e) => {
    setIsJSONEditorOpen(false);
    if (typeof e.json === 'string' && isValidJson(e.json)) {
      onChange(id, JSON.parse(e.json));
      return;
    }
    onChange(id, e.json);
  };

  const handleChangeSelectedContentType = (e, type) => {
    setCurrentContentType(type);
  };

  const handleOnDropdownChange = (e, data, currentField, message) => {
    const states = data?.value?.length > 0 ? data?.value : undefined;
    const statesToAdd = states?.map((state) => state.toLowerCase());
    const localCopyOfValue = _.cloneDeep(value);
    const currentContentTypeData = localCopyOfValue[currentContentType.id];

    if (!currentContentTypeData && data) {
      //Reference doesn't work with currentContentTypeData
      localCopyOfValue[currentContentType.id] = [
        createFieldRule(currentField, statesToAdd),
      ];
    } else if (
      doesPrefixExistInCurrentContentTypeData(
        currentContentTypeData,
        currentField,
      ) &&
      data
    ) {
      currentContentTypeData.push(createFieldRule(currentField, statesToAdd));
    } else {
      addNewStateToAlreadyExistingField(
        currentContentTypeData,
        currentField,
        statesToAdd,
        message,
      );
    }
    //The variable currentContentTypeData cannot be used here because of eslint and delete keyword
    if (localCopyOfValue[currentContentType.id]?.length === 0) {
      delete localCopyOfValue[currentContentType.id];
    }
    onChange(id, localCopyOfValue);
  };

  return (
    <>
      <div>
        {isJSONEditorOpen && (
          <ModalForm
            schema={JSONSchema(props)}
            onSubmit={onJSONSubmit}
            title={props.intl.formatMessage(messages.jsonTitle)}
            open={isJSONEditorOpen}
            formData={{ json: JSON.stringify(value, undefined, 2) }}
            onCancel={handleOnCancel}
            key="JSON"
          />
        )}
        <Container>
          <Button onClick={handleEditJSON} color="grey" id="json_button">
            <FormattedMessage id="Edit JSON" defaultMessage="Edit JSON" />
          </Button>
        </Container>
        <Divider />
      </div>
      <Sidebar.Pushable as={Segment} style={{ height: COMPONENT_HEIGHT }}>
        <SidebarComponent
          handleChangeSelectedContentType={handleChangeSelectedContentType}
          currentContentType={currentContentType}
          types={types}
          value={value}
        />
        <Sidebar.Pusher style={{ width: `calc(100% - ${SIDEBAR_WIDTH})` }}>
          <EditDataComponent
            request={request}
            handleOnDropdownChange={handleOnDropdownChange}
            currentContentType={currentContentType}
            value={value}
          />
        </Sidebar.Pusher>
      </Sidebar.Pushable>
    </>
  );
};
/**
 * Property types.
 * @property {Object} propTypes Property types.
 * @static
 */
VisualJSONWidget.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  description: PropTypes.string,
  required: PropTypes.bool,
  error: PropTypes.arrayOf(PropTypes.string),
  value: PropTypes.object,
  onChange: PropTypes.func,
  onEdit: PropTypes.func,
  onDelete: PropTypes.func,
  wrapped: PropTypes.bool,
  placeholder: PropTypes.string,
};

/**
 * Default properties.
 * @property {Object} defaultProps Default properties.
 * @static
 */
VisualJSONWidget.defaultProps = {
  description: null,
  required: false,
  error: [],
  value: null,
  onChange: null,
  onEdit: null,
  onDelete: null,
  title: '',
  id: '',
};
export default injectIntl(VisualJSONWidget);
