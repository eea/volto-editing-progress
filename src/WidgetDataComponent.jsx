import React, { useEffect, useState } from 'react';
import { Segment, Dropdown, Accordion, Icon } from 'semantic-ui-react';
import { flattenToAppURL } from '@plone/volto/helpers';
import { useSelector, useDispatch } from 'react-redux';
import { getRawContent } from './actions';
import { COMPONENT_HEIGHT } from './VisualJSONWidget';

import './less/editor.less';

export function makeFirstLetterCapital(string) {
  return string.charAt(0).toUpperCase() + string.slice(1);
}

const EditDataComponent = ({
  request,
  handleOnDropdownChange,
  currentContentType,
  value,
  fields,
  getDropdownValues,
}) => {
  const path = flattenToAppURL(
    '/@vocabularies/plone.app.vocabularies.WorkflowStates',
  );

  const dispatch = useDispatch();
  const requestStateOptions = useSelector((state) => state.rawdata?.[path]);
  const content = requestStateOptions?.data;

  useEffect(() => {
    if (
      path &&
      !requestStateOptions?.loading &&
      !requestStateOptions?.loaded &&
      !content
    )
      dispatch(getRawContent(path));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    dispatch,
    path,
    content,
    // eslint-disable-next-line react-hooks/exhaustive-deps
    requestStateOptions?.loaded,
    // eslint-disable-next-line react-hooks/exhaustive-deps
    requestStateOptions?.loading,
  ]);

  //Returns the saved values for dropdown with the first letter in uppercase

  const getValues = (currentField) => {
    if (
      !request.loading &&
      request.loaded &&
      currentContentType &&
      value[currentContentType.id]
    )
      return value[currentContentType.id].find(
        (rule) => rule?.prefix === currentField,
      );

    return undefined;
  };

  const renderLabel = (label) => ({
    color: 'blue',
    content: `${label.text}`,
  });

  const createStateOption = (stateOptions) => {
    return ['all', ...(stateOptions || [])].map((state) => ({
      key: makeFirstLetterCapital(state),
      text: makeFirstLetterCapital(state),
      value: makeFirstLetterCapital(state),
    }));
  };
  const [activeIndex, setActiveIndex] = useState(0);
  // const inputRef = useRef();
  const [inputValue, setInputValue] = useState('');
  const [conditionValue, setConditionValue] = useState('');
  const [linkValue, setLinkValue] = useState();
  const handleClick = (e, titleProps, currentField) => {
    const { index } = titleProps;
    const newIndex = activeIndex === index ? -1 : index;

    setActiveIndex(newIndex);
    setInputValue(getValues(currentField)?.linkLabel || '');
    setConditionValue(getValues(currentField)?.condition || '');
    setLinkValue(getValues(currentField)?.link || '');
  };
  const handleInputChange = (e, currentField) => {
    setInputValue(e.target.value);
    handleOnDropdownChange(
      null,
      { value: getValues(currentField).states || [] },
      currentField,
      e.target.value,
    );
  };
  const handleInputLinkChange = (e, currentField) => {
    setLinkValue(e.target.value);
    handleOnDropdownChange(
      null,
      { value: getValues(currentField).states || [] },
      currentField,
      undefined,
      undefined,
      e.target.value,
    );
  };
  const handleInputConditionChange = (e, currentField) => {
    setConditionValue(e.target.value);
    handleOnDropdownChange(
      null,
      { value: getValues(currentField).states || [] },
      currentField,
      undefined,
      e.target.value,
    );
  };
  return (
    <Segment
      style={{
        width: '100%',
        paddingBottom:
          request?.data?.fieldsets[0]?.fields.length > 9 ? '120px' : '',
        height: COMPONENT_HEIGHT,
        overflow: 'auto',
      }}
    >
      <Accordion styled fluid>
        {request?.loaded &&
          !request?.loading &&
          requestStateOptions?.loaded &&
          !requestStateOptions?.loading &&
          requestStateOptions?.data &&
          fields.map((currentField, index) => {
            if (
              request.data.required.includes(currentField) ||
              getDropdownValues(currentField) === undefined
            )
              return null;
            return (
              <React.Fragment key={`${currentField}${index}`}>
                <Accordion.Title
                  active={activeIndex === index}
                  index={index}
                  onClick={(e, titleProps) =>
                    handleClick(e, titleProps, currentField)
                  }
                  id={`property_${currentField}`}
                >
                  <div className="title-editing-progress">
                    <Icon name="dropdown" size="tiny" />
                    &nbsp;
                    {currentField}
                  </div>
                  <div className="title-editing-progress">
                    <Icon
                      name="cancel"
                      size="mini"
                      onClick={(e) => {
                        e.preventDefault();
                        handleOnDropdownChange(
                          e,
                          { value: undefined },
                          currentField,
                        );
                      }}
                    />
                  </div>
                </Accordion.Title>
                <Accordion.Content
                  active={activeIndex === index}
                  id={`property_content_${currentField}`}
                >
                  <label
                    htmlFor="message"
                    style={{ display: 'block', padding: '10px' }}
                  >
                    Message
                  </label>
                  <input
                    className="message-input"
                    value={inputValue}
                    onChange={(e) => handleInputChange(e, currentField)}
                    // ref={inputRef}
                    name="message"
                    style={{ padding: '10px' }}
                    disabled={getDropdownValues(currentField) == null}
                    placeholder="Write a dfferent message after you set at lest one state"
                  />
                  <label
                    htmlFor="message"
                    style={{ display: 'block', padding: '10px' }}
                  >
                    Link
                  </label>
                  <input
                    className="link-input"
                    value={linkValue}
                    onChange={(e) => handleInputLinkChange(e, currentField)}
                    // ref={inputRef}
                    name="link"
                    style={{ padding: '10px' }}
                    disabled={getDropdownValues(currentField) == null}
                    placeholder="Write a dfferent href link"
                  />
                  <label
                    htmlFor="condition"
                    style={{ display: 'block', padding: '10px' }}
                  >
                    Condition
                  </label>
                  <input
                    className="condition-input"
                    value={conditionValue}
                    onChange={(e) =>
                      handleInputConditionChange(e, currentField)
                    }
                    // ref={inputRef}
                    name="condition"
                    style={{ padding: '10px' }}
                    disabled={getDropdownValues(currentField) == null}
                    placeholder="Write a dfferent condition"
                  />
                  <label
                    htmlFor="dropdown"
                    style={{ display: 'block', padding: '10px' }}
                  >
                    States
                  </label>
                  <Dropdown
                    placeholder="Fields"
                    multiple
                    floating
                    selection
                    search
                    name="dropdown"
                    defaultValue={getDropdownValues(currentField)}
                    options={[
                      ...createStateOption(
                        requestStateOptions.data.items.map(
                          (option) => option.token,
                        ),
                      ),
                    ]}
                    onChange={(e, data) =>
                      handleOnDropdownChange(e, data, currentField)
                    }
                    renderLabel={renderLabel}
                  />
                </Accordion.Content>
              </React.Fragment>
            );
          })}
      </Accordion>
    </Segment>
  );
};

export default EditDataComponent;
