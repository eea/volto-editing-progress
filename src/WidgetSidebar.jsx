import React, { useEffect, useState } from 'react';
import { Sidebar, List } from 'semantic-ui-react';
import { SIDEBAR_WIDTH } from './VisualJSONWidget';

export const backgroundColor = (currentContentType, modified) => {
  let color = undefined;
  if (modified) {
    color = 'lightpink';
  }
  if (currentContentType) {
    color = 'lightblue';
  }
  return color;
};

const SidebarComponent = (props) => {
  const { types, currentContentType, handleChangeSelectedContentType } = props;
  // console.log(types);
  const [filtredTypes, setFiltredTypes] = useState({ ...types });
  const [inputValue, setInputValue] = useState('');

  useEffect(() => {
    setFiltredTypes({ ...types });
  }, [types]);

  const handleInputChange = (e) => {
    if (e.target.value == null) return;
    setInputValue(e.target.value);
    if (
      types.loaded &&
      !types.loading &&
      Array.isArray(types.types) &&
      types.types.length > 0
    ) {
      setFiltredTypes({
        ...types,
        types: types.types.filter((type) =>
          type.title.toLowerCase().includes(e.target.value.toLowerCase()),
        ),
      });
    }
  };

  return (
    <Sidebar
      as={List}
      animation="push"
      icon="labeled"
      visible
      relaxed
      size="big"
      divided
      selection
      style={{ width: SIDEBAR_WIDTH }}
    >
      <input
        value={inputValue}
        onChange={handleInputChange}
        placeholder="Search... "
        style={{ paddingLeft: ' 10px' }}
      />
      {filtredTypes.loaded &&
        !filtredTypes.loading &&
        Array.isArray(filtredTypes.types) &&
        filtredTypes.types.map((type) => (
          <List.Item
            style={{
              padding: '25px 5px',
              textAlign: 'center',
              backgroundColor: backgroundColor(
                currentContentType?.id === type.id,
                Object.keys(props.value).includes(type.id),
              ),
            }}
            key={type.id}
            id={`sidebar_${type.id}`}
            onClick={(e) => handleChangeSelectedContentType(e, type)}
          >
            <List.Content>
              <List.Header>{type.title}</List.Header>
            </List.Content>
          </List.Item>
        ))}
    </Sidebar>
  );
};

export default SidebarComponent;
