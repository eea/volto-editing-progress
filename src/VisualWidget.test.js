import { makeFirstLetterCapital } from './WidgetDataComponent';
import React from 'react';
import { Provider } from 'react-intl-redux';
import { MemoryRouter } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { backgroundColor } from './WidgetSidebar';
import '@testing-library/jest-dom/extend-expect';

import VisualJSONWidget from './VisualJSONWidget';

const mockStore = configureStore();
const propsEmpty = {};

describe('Widget Data Component', () => {
  it('should make first letter capital', () => {
    const testString = 'this is a test string';
    expect(makeFirstLetterCapital(testString)[0]).toEqual(
      testString[0].toUpperCase(),
    );
  });
});

describe('Widget Sidebar', () => {
  it('should return a background lightblue', () => {
    expect(backgroundColor(true, false)).toEqual('lightblue');
    expect(backgroundColor(true, true)).toEqual('lightblue');
  });
  it('should return a background lightpink', () => {
    expect(backgroundColor(false, true)).toEqual('lightpink');
  });
});

describe('Visual widget', () => {
  it('renders the VisualJSONWidget component without breaking if props and progressEditing are empty', () => {
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
      progressEditing: {},
      rawdata: {},
      types: {
        loaded: true,
        loading: false,
        types: [
          {
            id: 'content-type-1',
            title: 'Content Type 1',
            '@id': '/content-type-1',
          },
        ],
      },
    });
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <VisualJSONWidget
            pathname="/test"
            {...propsEmpty}
            hasToolbar={true}
            value={{}}
            onChange={jest.fn()}
          />
        </MemoryRouter>
      </Provider>,
    );

    expect(screen.getByText('Edit JSON')).toBeInTheDocument();
    expect(screen.getByText('Add Property')).toBeInTheDocument();
    expect(
      container.querySelector('input[placeholder="Search... "]'),
    ).toBeInTheDocument();
    expect(screen.getByText('Content Type 1')).toBeInTheDocument();
  });
});
