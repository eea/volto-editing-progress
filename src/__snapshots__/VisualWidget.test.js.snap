// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visual widget renders the VisualJSONWidget component without breaking if props and progressEditing are empty 1`] = `
Array [
  <div>
    <div
      className="ui container"
    >
      <button
        className="ui grey button"
        id="json_button"
        onClick={[Function]}
      >
        Edit JSON
      </button>
      <div
        aria-expanded={false}
        className="ui dropdown ui grey button dropdown-button"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="listbox"
        tabIndex={0}
      >
        <div
          aria-atomic={true}
          aria-live="polite"
          className="divider text"
          role="alert"
        >
          Add Property
        </div>
        <i
          aria-hidden="true"
          className="dropdown icon"
          onClick={[Function]}
        />
        <div
          className="menu transition"
        />
      </div>
    </div>
    <div
      className="ui divider"
    />
  </div>,
  <div
    className="ui segment pushable"
    style={
      Object {
        "height": "750px",
      }
    }
  >
    <div
      className="ui big divided selection relaxed list ui push left visible sidebar"
      icon="labeled"
      role="list"
      style={
        Object {
          "width": "250px",
        }
      }
    >
      <input
        onChange={[Function]}
        placeholder="Search... "
        style={
          Object {
            "paddingLeft": " 10px",
          }
        }
        value=""
      />
    </div>
    <div
      className="pusher"
      style={
        Object {
          "width": "calc(100% - 250px)",
        }
      }
    >
      <div
        className="ui segment"
        style={
          Object {
            "height": "750px",
            "overflow": "auto",
            "paddingBottom": "",
            "width": "100%",
          }
        }
      >
        <div
          className="accordion ui fluid styled"
        />
      </div>
    </div>
  </div>,
]
`;
